src/Interpreters/ComparisonTupleEliminationVisitor.cpp

```diff
@@ -0,0 +1,108 @@
+#include <Interpreters/ComparisonTupleEliminationVisitor.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTLiteral.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+namespace
+{
+
+
+ASTs splitTuple(const ASTPtr & node)
+{
+    if (const auto * func = node->as<ASTFunction>(); func && func->name == "tuple")
+        return func->arguments->children;
+
+    if (const auto * literal = node->as<ASTLiteral>(); literal && literal->value.getType() == Field::Types::Tuple)
+    {
+        ASTs result;
+        const auto & tuple = literal->value.get<const Tuple &>();
+        for (const auto & child : tuple)
+            result.emplace_back(std::make_shared<ASTLiteral>(child));
+        return result;
+    }
+
+    return {};
+}
+
+
+ASTPtr concatWithAnd(const ASTs & nodes)
+{
+    if (nodes.empty())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot concat empty list of nodes");
+
+    if (nodes.size() == 1)
+        return nodes[0];
+
+    auto result = makeASTFunction("and");
+    result->arguments->children = nodes;
+    return result;
+}
+
+void trySplitTupleComparsion(ASTPtr & expression)
+{
+    if (!expression)
+        return;
+
+    auto * func = expression->as<ASTFunction>();
+    if (!func)
+        return;
+
+    if (func->name == "and" || func->name == "or" || func->name == "not" || func->name == "tuple")
```

Comment:
Not sure why we dive into `tuple` function, can you please provide example of use-case ?